### DISCUSSION TITLE
Player Release 1.10@next: It’s Time for the Big One (Jan 2024)

### DISCUSSION THREAD

**Discussion by mihar-22** [vidstack/player#1116] [Releases]
👋

Hey friends!

This is our final feature release before the official 1.0! It's been an enormous amount of work and it's taken a while but it's well worth the wait. This release was focused on improving DX, adding missing core features, ironing out tons of bugs, and just making the library feel more ready for prime time. I'm sure there's new bugs and issues thanks to features/changes that we haven't caught yet; we'd love your help in opening descriptive issues with repros so we can get them all fixed! Okay... let's dig in!

✨ What's New?

- Bundler Plugins
- Plyr Layout
- New Default Audio Layout
- Layout Slots
- Font Customization Menu
- Keyboard Action Display
- New Load Strategies
- Clipping
- Storage
- Remote Playback
- Media Session
- Text Tracks Content
- New Thumbnail Source Types
- New Tailwind Variants
- hls.js 1.5
- Tons of bug fixes!

## Breaking Changes

**Player Queries**

We've removed player queries as they relied on eval and posed a security risk. We've replaced them with `MediaPlayerQuery` subscriptions. See the updated [default layout sizes][default-layout-sizes] guide for how to use them. This change also affects the `media-layout` element, see the updated [layouts][media-layout] page.

**Naming Changes**

We've fixed some prop/callback names (not attributes) to be consistently camel-cased. This isn't a breaking change yet but it might be when you upgrade to 1.0.

Prop name changes:

- `crossorigin` -> `crossOrigin`
- `playsinline` -> `playsInline`
- `autoplay` -> `autoPlay`

Callback name changes:

- `onAutoplay` -> `onAutoPlay`
- `onAutoplayFail` -> `onAutoPlayFail`

[default-layout-sizes]: https://vidstack.io/docs/player/components/layouts/default-layout#size
[media-layout]: https://vidstack.io/docs/wc/player/styling/responsive-design#layouts

## Bundler Plugins

🚨 **This feature is only for Web Component users.**

We now have bundler plugins coming that work with Astro, Nuxt, SvelteKit, SolidStart, and anywhere that uses Vite/Webpack. The plugin was built to help with auto-importing the components and styles you use. No more manual setup required and load only what you use!

```ts
// Once plugin is setup, this is all you need to import!
import 'vidstack/bundle';
```

<img width="600px" alt="Installation page showing new bundler section with Astro, SvelteKit, Vite, Webpack, and None." src="https://github.com/vidstack/player/assets/14304599/fb7e199a-dadd-4df9-81ec-1f1e07f2654a">

📝 [Available in our installation docs.][vds-installation]

📹 [Click here for video explainer.][bundler-plugins]

[vds-installation]: https://vidstack.io/docs/player/getting-started/installation
[bundler-plugins]: https://twitter.com/mihar_22/status/1746789001439457448

## Plyr Layout

The beautiful Plyr Layout, designed by [Sam Potts][sam-potts] all the way back in 2015, has been newly rebuilt! We've included the same features, styles, CSS variables, icons, and relevant options to make migrating over simpler.

<img width="100%" alt="The beautiful ocean themed Plyr video layout." src="https://github.com/vidstack/player/assets/14304599/2f0f7a00-0da9-46f1-bece-c44483577afb">

📝 [Available in our installation docs.][vds-installation]

📝 [New Plyr Layout component page][plyr-layout]

[sam-potts]: https://twitter.com/sam_potts
[plyr-layout]: https://vidstack.io/docs/player/components/layouts/plyr-layout

## New Default Audio Layout

[Billy Gözütok][billy] and I worked together to redesign our Default Audio Layout. We wanted it to be as compact as possible and optimized for touch. We mainly did this because the previous design was pretty clunky/ugly, there was a big height difference between small/large layouts, and most importantly it wasn't adapting well with the new load play strategy. Let us know what you think!

https://github.com/vidstack/player/assets/14304599/4b5cd1bf-11bf-4ea4-9c48-4f149bf2533b

> The video above was an early preview, transitions have been further refined.

[billy]: https://twitter.com/billygozutok
[new-audio-layout]: https://twitter.com/mihar_22/status/1749971530879258640

## Layout Slots

🚨 **This feature is only for React users.**

Slots are now available on all layouts for inserting or replacing content inside of them. This solves the long issue of, "I only want to add/change a few things, I don't want to build my own layout."

```tsx
<PlyrLayout
  slots={{
    beforePlayButton: null,
    playButton: CustomPlayButton,
    afterPlayButton: null,
    /** 57 other slots positions... */
  }}
/>
```

```tsx
<DefaultVideoLayout
  slots={{
    afterPlayButton: null,
    playButton: CustomPlayButton,
    beforePlayButton: null,
    /** 72 other slots positions... */
  }}
/>
```

📝 [Plyr Layout slots docs.][plyr-layout-slots]

📝 [Default Layout slots docs.][default-layout-slots]

[default-layout-slots]: https://vidstack.io/docs/player/components/layouts/default-layout#slots
[plyr-layout-slots]: https://vidstack.io/docs/player/components/layouts/plyr-layout#slots

## Font Customization Menu

The default audio and video layouts now have a font customization menu to improve the accessibility of captions.

> User font settings will override your captions default theme CSS variables.

https://github.com/vidstack/player/assets/14304599/697f23b8-1115-4bdd-ac63-0235437f37de

## Keyboard Action Display

Similar to YouTube Player, a keyboard action display has landed to improve accessibility for our default video layout. It displays the keyboard action as an icon and any relevant text on screen. It also includes an ARIA status update, i18n, and custom icons.

https://github.com/vidstack/player/assets/14304599/921e884f-0658-41ba-9c86-ac90826d5809

## New Load Strategies

There is a new load strategy called `play`. This new option will delay loading _both_ the provider and media until a play request is made. This works best when you want to delay loading until interaction.

Secondly, we've separated the poster loading strategy into `posterLoad` with similar options. Poster loading has been separated so you can display an image before media is ready for playback. This generally works well in combination with load="play" to create previews.

```html
<media-player load="play" poster-load="visible" />
```

```tsx
<MediaPlayer load="play" posterLoad="visible" />
```

📝 [Updated load strategies docs.][load-strategies]

[load-strategies]: https://vidstack.io/docs/player/core-concepts/loading#load-strategies

## Clipping

Clipping allows shortening the media by specifying the time at which playback should start and end. The media and any resources such as chapters durations are updated to match the new clipped length. In addition, [Media URI Fragments][media-uri-frags] are used internally to efficiently load audio and video files between the clipped start and end times (e.g., /video.mp4#t=30,60).

```html
<media-player clip-start-time="30" clip-end-time="60"></media-player>
```

```tsx
<MediaPlayer clipStartTime={30} clipEndTime={60}></MediaPlayer>
```

📝 [Clipping docs.][clipping]

[clipping]: https://vidstack.io/docs/player/core-concepts/loading#clipping
[media-uri-frags]: https://www.w3.org/TR/media-frags

## Storage

Storage enables saving player and media settings so that the user can resume where they left off. This includes saving and initializing on load settings such as language, volume, muted, captions visibility, and playback time. The player supports both local and remote storage.

📝 [Storage docs.][storage]

[storage]: https://vidstack.io/docs/player/core-concepts/loading#storage

## Remote Playback

The player can now cast to a remote playback device such as a TV using AirPlay or Google Cast. The entire process is async and lazy loaded to avoid bloating the library and your application. The buttons are included when supported in the Default Video Layout.

📝 [Remote Playback API docs.][remote-playback-api]

📝 [AirPlay Button][airplay-button]

📝 [Google Cast Button][google-cast-button]

📝 [Google Cast Provider][google-cast-provider]

📹 [Click here to see Google Cast demo.][google-cast-demo]

[remote-playback-api]: https://vidstack.io/docs/player/api/remote-playback
[airplay-button]: https://vidstack.io/docs/player/components/buttons/airplay-button
[google-cast-button]: https://vidstack.io/docs/player/components/buttons/google-cast-button
[google-cast-provider]: https://vidstack.io/docs/player/api/providers/google-cast
[google-cast-demo]: https://twitter.com/mihar_22/status/1744167651847586120

## Media Session

The [Media Session API][media-session-api] is now automatically set for you so media notifications and handlers are displayed on your device/browser. There's also a new `artist` player property for setting the artist metadata (in the example below it's "Blender Foundation").

<img width="400px" alt="Installation page showing new bundler section with Astro, SvelteKit, Vite, Webpack, and None." src="https://github.com/vidstack/player/assets/14304599/d4d7cf54-1661-4559-a39d-213bd722fd91">

[media-session-api]: https://developer.mozilla.org/en-US/docs/Web/API/Media_Session_API

## Text Tracks Content

You can now pass serialized or deserialized JSON content directly into the `TextTrack` object or `<Track>` component.

```tsx
import { type VTTContent } from '@vidstack/react';

const content: VTTContent = {
  regions: [],
  cues: [
    { startTime: 0, endTime: 5, text: '...' },
    { startTime: 5, endTime: 10, text: '...' },
  ],
};

// Provide JSON directly (can also be serailized - JSON.stringify).
<Track content={content} label="English" kind="captions" lang="en-US" type="json" />;
```

📝 [Updated loading text tracks docs.][text-tracks]

[text-tracks]: https://vidstack.io/docs/player/core-concepts/loading#text-tracks

## Thumbnail Source Types

We've expanded support to a variety of thumbnail source types:

- VTT (sprite or multiple images).
- JSON (VTT cues, array of images, or storyboard).
- Objects.
- [Mux storyboards][mux-storyboards].

**Storyboard Example (JSON)**

```json
{
  "url": "https://example.com/storyboard.jpg",
  "tileWidth": 256,
  "tileHeight": 160,
  "tiles": [
    { "startTime": 0, "x": 0, "y": 0 },
    { "startTime": 50, "x": 256, "y": 0 }
  ]
}
```

```html
<media-video-layout thumbnails="/storyboard.json" />
```

```tsx
<DefaultVideoLayout thumbnails="/storyboard.json" />
```

**Storyboard Example (Object)**

```ts
import { type ThumbnailStoryboard } from 'vidstack';

const storyboard: ThumbnailStoryboard = {
  url: 'https://example.com/storyboard.jpg',
  tileWidth: 256,
  tileHeight: 160,
  tiles: [
    { startTime: 0, x: 0, y: 0 },
    { startTime: 50, x: 256, y: 0 },
  ],
};
```

Can be provided directly to any of the following:

<details>
 <summary>Web Components</summary>

```ts
// 1. Layouts.
const videoLayout = document.querySelector('media-video-layout');
videoLayout.thumbnails = storyboard;

// 2. Thumbnail component.
const thumbnail = document.querySelector('media-thumbnail');
thumbnail.src = storyboard;
```

</details>

<details>
 <summary>React</summary>

```tsx
// 1. Layouts.
<PlyrLayout thumbnails={storyboard} />
<DefaultVideoLayout thumbnails={storyboard} />

// 2. Thumbnail component.
<Thumbnail.Root src={storyboard}>...</Thumbnail.Root>

// 3. Hook
const thumbnails = useThumbnails(storyboard);

```

</details>

<br />

📝 [Updated loading thumbnails docs.][thumbnails]

📝 [Updated `useThumbnails` docs.][thumbnails]

[thumbnails]: https://vidstack.io/docs/player/core-concepts/loading#thumbnails
[use-thumbnails]: https://vidstack.io/docs/player/api/hooks/use-thumbnails
[mux-storyboards]: https://docs.mux.com/guides/create-timeline-hover-previews#json

## New Tailwind Variants

There's heaps of new Tailwind variants for view type, stream type, and remote playback.

```html
<div class="media-view-audio:... media-stream-dvr:..." />
```

📝 [Updated Tailwind variants docs.][tw-variants]

[tw-variants]: https://vidstack.io/docs/player/styling/tailwind#media-variants

## hls.js 1.5

The latest hls.js release (v1.5) takes advantage of [ManagedMediaSource (MMS)][mms] to finally run on iPhone. We've now included it in our support checks so hls.js will be used as the default streaming engine across all modern browsers starting from iOS 17+. This is a huge win for the web as we can use a single and consistent configurable streaming engine. If you've ever dealt with DRM or setting headers then this will be a huge blessing. It also means playback quality options are available and can be set on iPhone finally.

[mms]: https://github.com/w3c/media-source/issues/320

## Misc.

- Switching sources on our CDN bundle was broken.
- New `crossOrigin` property on poster, slider video, and thumbnail components.
- New `disabled` prop on gesture components.
- New default layout props: `noGestures`, `seekStep`, and `sliderChaptersMinWidth`.
- Slider chapters in the default layout are now shown based on the width of the slider itself and not the player.
- Media request callbacks are now called on all components and can be cancelled. [See cancelling requests][cancelling-media-requests].
- New `<Title>` component.
- New `duration` player property.
- New hooks: `useChapterTitle` and `useTextCues`.
- Fixed white border appearing around posters and thumbnails due to missing `src` attribute.
- Fixed tooltips occasionally freezing when buttons were clicked.
- Menu resizing was fixed.
- Vimeo chapters are now loaded from the embed.
- Tons of other bug fixes and improvements, [see changelog][changelog].

[changelog]: https://github.com/vidstack/player/blob/main/CHANGELOG.md
[cancelling-media-requests]: https://vidstack.io/docs/player/core-concepts/events#cancelling-requests

## What's Next?

It's been an incredible 3+ long years of building media players and this library. For the first time, we're in an amazing place and I can say we're on the brink of being 1.0-ready. We're technically done with building new features but knowing me I might sneak in a suprise or two. From here, I'll be addressing new bugs as they roll in and working on our player product page to help us finally market it. After the official 1.0 announcement, we'll start work on our CMS and Next.js integration around mid-Feb and start rolling it out to users on the waitlist (as seen on our home page). It's honestly a busy time for me as we're doing a bunch of house work and I have a newborn son on the way in April, so I'll try my best to balance everything out!

In conclusion, with ManagedMediaSource and the Fullscreen API landing on iOS, creating custom media players across the web is becoming a reality. We believe that the web, in combination with React and React Native, is emerging as the de facto platform for video to thrive. We are thrilled to be at the forefront of building customizable media experiences on the frontend. We're incredibly excited for our future libraries and services that will follow the player! Onwards!

Until next time friends 🍎

**Comment by BlackDahlia313** [comment]
Out of the box Nuxt support?!? Whoah! This is amazing news! Great work! What an amazing release!

**Comment by thomaskant** [comment]
Awesome work! 🚀✨

**Comment by Jumaron** [comment]
Awesome! Tbh adding simple VAST Prerole support like Plyr has with an ad.url would make Plyr for a lot of people I know myself included obsolete. I really love the Design behind vidstack but thats the only thing its missing/I'am waiting for since it forces me to stick with Plyr ^^
So keep going! :)

**Reply by shmaltz** [reply]
This. Please add VAST pre roll with ad.url support.
Thank you 😊
